qkeras/quantizers.py

```diff
@@ -439,9 +443,517 @@ def trainable_variables(self):
+  def __init__(
+      self,
+      bits=8,
+      integer=0,
+      symmetric=1,
+      keep_negative=True,
+      alpha=None,
+      use_stochastic_rounding=False,
+      scale_axis=None,
+      qnoise_factor=1.0,
+      var_name=None,
+      use_variables=False,
+  ):
+    super(quantized_linear, self).__init__()
+
+    # Set _initialized parameter to False to prevent the setters from
+    # performing preliminary calculations
+    self._initialized = False
+    self.var_name = var_name
+    self.bits = bits
+    self.integer = integer
+    self.symmetric = symmetric
+    self.keep_negative = keep_negative
+    self.qnoise_factor = qnoise_factor
+    self.use_stochastic_rounding = use_stochastic_rounding
+    self.alpha = alpha
+    self.scale_axis = scale_axis
+    self.use_variables = use_variables
+
+    # Perform preliminary calculations based on attributes above
+    self._initialized = True
+    self._set_default_quantization_scale()
+
+
+  @property
+  def bits(self):
+    return self._bits
+
+  @bits.setter
+  def bits(self, bits):
+    self._bits = bits
+    err_msg = f"Bit count {bits} must be positive"
+    if bits <= 0:
+      raise ValueError(err_msg)
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def integer(self):
+    return self._integer
+
+  @integer.setter
+  def integer(self, integer):
+    self._integer = integer
+    err_msg = (f"Integer bit count {integer} must be nonnegative")
+    if integer < 0:
+      raise ValueError(err_msg)
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def symmetric(self):
+    return self._symmetric
+
+  @symmetric.setter
+  def symmetric(self, symmetric):
+    self._symmetric = symmetric
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def keep_negative(self):
+    return self._keep_negative
+
+  @keep_negative.setter
+  def keep_negative(self, keep_negative):
+    self._keep_negative = keep_negative
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def use_stochastic_rounding(self):
+    return self._use_stochastic_rounding
+
+  @use_stochastic_rounding.setter
+  def use_stochastic_rounding(self, use_stochastic_rounding):
+    self._use_stochastic_rounding = use_stochastic_rounding
+
+  @property
+  def alpha(self):
+    return self._alpha
+
+  @alpha.setter
+  def alpha(self, alpha):
+    """Set alpha and check if it isvalid."""
+
+    self._alpha = alpha
+
+    # Error checking
+    if isinstance(alpha, six.string_types):
+      # Check the quantizer has been given a valid alpha string
+      if not alpha in self.ALPHA_OPTIONS:
+        raise ValueError(
+            f"Invalid alpha '{alpha}' for auto alpha computation. "
+            f"Must be one of {self.ALPHA_OPTIONS}")
+    elif alpha is not None: # alpha is a tensor
+      try:
+        # any allowable array type can be cast as a numpy array
+        np.array(alpha)
```

Comment:
in original design, alpha =(1.0, "auto", "auto_po2", None) are the only viable options. it doesn't allow the direct setting of scale value through alpha except when alpha=1.0, which means no custom scaling is allowed. I will argue by setting scale here through alpha it might be confusing because integer_bits in the quantizer will interact with alpha when calculating the overall quantization scale. Therefore it won't make sense for users to specify scale value through alpha since it won't be the final quantization scale anyway and it will only be confusing.

Reply:
The three different scales are confusing, and we should definitely discuss this more. If we wanted to simplify this, I believe it would be best to use a single scale, the quantization scale, and then rework qtools (and anything else that uses the `.scale` attribute) to handle this properly.

As for allowing the user to set the scale as a tensor, there are a couple of use cases I had in mind. 
1. Treating the scale like a hyperparameter, as in pokeBNN
2. Computing the scale on the training data, and then "freezing" it on the test data.

Reply:
I agree with Matt's points. Fixing the scale will be a very useful option in the near future so we can reproduce PokeBNN and other work.